bjs2 26872 Добавление аватара пользователя

Alginin opened the pull request.

Alginin committed:
Добавлены методы добавления и удаления

Alginin committed:
Добавлены метод получения изображения

Alginin committed:
Реализовал методы:
1.Сохранение в облочное хранилище аватара (большое и маленькое изображение)  с применением сжатия при необходимости.
2.Получение аватара пользователя.
3.Удаления аватара пользователя из облочного хранилища и базы данных.

mrkorolev commented on src/main/java/school/faang/user_service/config/s3/S3Config.java:
```diff
@@ -0,0 +1,33 @@
+        BasicAWSCredentials credentials = new BasicAWSCredentials(accessKey, secretKey);
+        return AmazonS3ClientBuilder
+                .standard()
+                .withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(endpoint, region))
```
👍

mrkorolev commented on src/main/java/school/faang/user_service/controller/user/UserController.java:
```diff
@@ -38,4 +46,42 @@ public UserDto getUser(@PathVariable @Valid long userId) {
+            imageBytes = userService.getBigImageFromProfile(userId).readAllBytes();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
```
Вот эту штуку лучше внутри сервиса делать.

mrkorolev commented on src/main/java/school/faang/user_service/controller/user/UserController.java:
```diff
@@ -38,4 +46,42 @@ public UserDto getUser(@PathVariable @Valid long userId) {
+
+    @GetMapping("/{user-id}/profile-small-image")
+    public ResponseEntity<byte[]> getSmallImageFromProfile(@PathVariable("user-id") Long userId) {
+        byte[] imageBytes = null;
+        try {
+            imageBytes = userService.getSmallImageFromProfile(userId).readAllBytes();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+        HttpHeaders headers = new HttpHeaders();
+        headers.setContentType(MediaType.IMAGE_JPEG);
+        return new ResponseEntity<>(imageBytes, headers, HttpStatus.OK);
```
Тот же самый метод, что выше - отличается только url.

20cognosce replied:
Дублирование кода

mrkorolev commented on src/main/java/school/faang/user_service/service/s3/S3Service.java:
```diff
@@ -0,0 +1,13 @@
+import java.io.InputStream;
+
+public interface S3Service {
+    void uploadFile(InputStream data, ObjectMetadata metadata, String folder, Long userId);
+
+    InputStream downloadFile(String fileKey);
+
+    void deleteFile(String fileKey);
```
У меня такой же - надеюсь не будет конфликтов.

Alginin committed:
Реализовал методы:
1.Сохранение в облочное хранилище аватара (большое и маленькое изображение)  с применением сжатия при необходимости.
2.Получение аватара пользователя.
3.Удаления аватара пользователя из облочного хранилища и базы данных.

mrkorolev commented on src/main/resources/application.yaml:
```diff
@@ -42,4 +47,18 @@ project-service:
+    bucketName: corpbucket
+    region: us-east-1
+
+reduce:
```
Продумай нейминг - image-limitations, или image

mrkorolev commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -29,14 +48,118 @@ public List<User> findPremiumUser(UserFilterDto filterDto) {
+        s3Service.deleteFile(large);
+        s3Service.deleteFile(compressed);
+
+        user.setUserProfilePic(null);
+        userRepository.save(user);
```
Тут мне кажется ошибка - должна быть дефолтная ава, из моей задачи (вместе подумаем, как это сделать).

20cognosce replied:
Дефолтная ава - да, это которую Никита генерирует из внешнего сервиса

mrkorolev reviewed (commented):
В целом ок, за исключением удаления файла - надо подумать, как будем обходить ситуацию с удалением файла (надо будет использовать дефолтную аву)

Alginin committed:
Добавил часть тестов

Alginin committed:
Добавил тестов

20cognosce commented on src/main/java/school/faang/user_service/controller/user/UserController.java:
```diff
@@ -38,4 +46,42 @@ public UserDto getUser(@PathVariable @Valid long userId) {
+        userService.addProfileImage(userId, file);
+    }
+
+    @PutMapping("/{user-id}/profile-image")
```
DeleteMapping

20cognosce commented on src/main/java/school/faang/user_service/controller/user/UserController.java:
```diff
@@ -38,4 +46,42 @@ public UserDto getUser(@PathVariable @Valid long userId) {
+        userService.deleteProfileImage(userId);
+    }
+
+    @GetMapping("/{user-id}/profile-image")
```
После id уже не делай subpath
используй query параметры или в теле запроса передавай
или перед айди путь поменяй

20cognosce commented on src/main/java/school/faang/user_service/controller/user/UserController.java:
```diff
@@ -38,4 +46,42 @@ public UserDto getUser(@PathVariable @Valid long userId) {
+
+    @GetMapping("/{user-id}/profile-image")
+    public ResponseEntity<byte[]> getBigImageFromProfile(@PathVariable("user-id") Long userId) {
+        byte[] imageBytes = null;
+        try {
+            imageBytes = userService.getBigImageFromProfile(userId).readAllBytes();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+        HttpHeaders headers = new HttpHeaders();
+        headers.setContentType(MediaType.IMAGE_JPEG);
+        return new ResponseEntity<>(imageBytes, headers, HttpStatus.OK);
```
Что это делает в контроллере
Почему байты отдаешь, отдавай Resource

20cognosce commented on src/main/java/school/faang/user_service/service/s3/S3ServiceImpl.java:
```diff
@@ -0,0 +1,38 @@
+@Service
+@RequiredArgsConstructor
+public class S3ServiceImpl implements S3Service {
+    private String bucketName;
```
@Value

20cognosce commented on src/main/java/school/faang/user_service/service/s3/S3ServiceImpl.java:
```diff
@@ -0,0 +1,38 @@
+    private String bucketName;
+    private final AmazonS3 s3Client;
+
+    @Autowired
+    public S3ServiceImpl(@Value("${services.s3.bucketName}") String bucketName, AmazonS3 s3Client) {
+        this.s3Client = s3Client;
+        this.bucketName = bucketName;
+    }
```
Убирай такой конструктор

20cognosce commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -29,14 +48,118 @@ public List<User> findPremiumUser(UserFilterDto filterDto) {
-}
+
+    @Transactional
+    public void addProfileImage(Long userId, MultipartFile file) throws IOException {
```
Что это за трешачок
В три раза короче должны быть методы

IO операции вынеси из транзакции, они их не поддерживают

20cognosce commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -29,14 +48,118 @@ public List<User> findPremiumUser(UserFilterDto filterDto) {
+        UserProfilePic userProfilePic = new UserProfilePic();
+
+        if (width <= largeImage && height <= largeImage) {
+            InputStream ImageStream = bufferedImageToInputStream(originalImage, "jpg");
```
Поток закрываем?

20cognosce commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -29,14 +48,118 @@ public List<User> findPremiumUser(UserFilterDto filterDto) {
+            InputStream ImageStream = bufferedImageToInputStream(originalImage, "jpg");
+            s3Service.uploadFile(ImageStream, metadata, filePath, userId);
+            userProfilePic.setFileId(filePath);
+        } else {
+            double scaleFactor = width > height ?
+                    (double) largeImage / width :
+                    (double) largeImage / height;
+
+            int targetWidth = (int) (width * scaleFactor);
+            int targetHeight = (int) (height * scaleFactor);
+
+            BufferedImage maxLargeImage = resizeImage(originalImage, targetWidth, targetHeight);
+            InputStream largeImageStream = bufferedImageToInputStream(maxLargeImage, "jpg");
+
+            s3Service.uploadFile(largeImageStream, metadata, filePath, userId);
+            userProfilePic.setFileId(filePath);
+        }
+
+        if (width > smallImage || height > smallImage) {
+
+            double scaleFactor = width > height ?
+                    (double) smallImage / width :
+                    (double) smallImage / height;
+
+            int targetWidth = (int) (width * scaleFactor);
+            int targetHeight = (int) (height * scaleFactor);
+
+            BufferedImage maxSmallImage = resizeImage(originalImage, targetWidth, targetHeight);
+            InputStream smallImageStream = bufferedImageToInputStream(maxSmallImage, "jpg");
+            String filePathSmallImage = filePath + "_small";
+
+            s3Service.uploadFile(smallImageStream, metadata, filePathSmallImage, userId);
+            userProfilePic.setSmallFileId(filePathSmallImage);
+        }
+        user.setUserProfilePic(userProfilePic);
+        userRepository.save(user);
```
Переделать всю эту порнографию

20cognosce commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -29,14 +48,118 @@ public List<User> findPremiumUser(UserFilterDto filterDto) {
+    }
+
+    @Transactional(readOnly = true)
+    public InputStream getBigImageFromProfile(Long userId) {
+        User user = userRepository.findById(userId).orElseThrow();
+        return s3Service.downloadFile(user.getUserProfilePic().getFileId());
+    }
+
+    @Transactional(readOnly = true)
+    public InputStream getSmallImageFromProfile(Long userId) {
```
Как ты можешь отдавать из метода поток ввода...

20cognosce replied:
Класс создай свой и отдавай объект с контентом в байтах
Resource используй
Multipart может

Но не это

20cognosce replied:
Ну и транзакции не работают на IO

20cognosce commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -29,14 +48,118 @@ public List<User> findPremiumUser(UserFilterDto filterDto) {
+
+    private InputStream bufferedImageToInputStream(BufferedImage image, String format) throws IOException {
+        ByteArrayOutputStream os = new ByteArrayOutputStream();
+        ImageIO.write(image, format, os);
```
Потоки закрывать надо

20cognosce commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -29,14 +48,118 @@ public List<User> findPremiumUser(UserFilterDto filterDto) {
+                Scalr.OP_ANTIALIAS);
+    }
+
+    private InputStream bufferedImageToInputStream(BufferedImage image, String format) throws IOException {
```
Не отдавать потоки, утечка памяти

20cognosce commented on src/main/resources/application.yaml:
```diff
@@ -42,4 +47,18 @@ project-service:
+  picture-allowed-types: image/jpeg, image/png
+  max-pixel-size:
+    large-image: 1080
+    small-image: 170
```
В какие ресурсы ты картинку положил
Не в тестовые нифига

20cognosce reviewed (changes requested):
Давай по новой

Alginin committed:
Частичное исправление замечаний по ревью

20cognosce commented on src/main/java/school/faang/user_service/config/redis/RedisConfig.java:
```diff
@@ -0,0 +1,45 @@
+    @Bean
+    public MessageListener userBanListenerAdapter() {
+        return new MessageListenerAdapter(userBanListener);
+    }
```
Что это всё делает в задаче добавления аватара?

20cognosce reviewed (changes requested):
Разберись с гитом

Alginin committed:
Решение проблемы с гит

Alginin committed:
Merge branch 'kraken-master-stream6' into feature-BJS2-26872

# Conflicts:
#	src/main/java/school/faang/user_service/controller/advice/ApiControllerAdvice.java
#	src/main/java/school/faang/user_service/controller/user/UserController.java
#	src/main/java/school/faang/user_service/service/user/UserService.java
#	src/main/resources/application.yaml
#	src/test/java/school/faang/user_service/service/skill/SkillServiceTest.java
#	src/test/java/school/faang/user_service/service/subscription/SubscriptionServiceTest.java
#	src/test/java/school/faang/user_service/service/user/UserServiceTest.java

Alginin committed:
Решение проблемы с гит №2

20cognosce commented on src/main/java/school/faang/user_service/controller/advice/ApiControllerAdvice.java:
```diff
@@ -0,0 +1,55 @@
+        );
+    }
+
+    // DTO (Jakarta) validation exceptions
```
Комментарии кому

20cognosce commented on src/main/java/school/faang/user_service/controller/advice/ApiControllerAdvice.java:
```diff
@@ -0,0 +1,55 @@
+        return new ErrorResponse(
+                HttpStatus.INTERNAL_SERVER_ERROR.value(),
+                e.getMessage()
+//                "Something wrong on our end, contact IT support immediately."
```
?

20cognosce commented on src/main/java/school/faang/user_service/service/s3/S3ServiceImpl.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    @Override
+    public InputStream downloadFile(String filePath) {
```
Не принимать и не отдавать потоки, нет
Работай с файлом

mrkorolev replied:
Погоди, я же определял интерфейс для этой темы, что там случилось?

20cognosce commented on src/main/java/school/faang/user_service/service/s3/S3ServiceImpl.java:
```diff
@@ -0,0 +1,32 @@
+    private final AmazonS3 s3Client;
+
+    @Override
+    public void uploadFile(InputStream data, ObjectMetadata metadata, String filePath, Long userId) {
```
data - не inputStream, нет

mrkorolev replied:
И тут, там вроде byte[] data

20cognosce commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -114,4 +137,101 @@ public User getUserById(Long userId) {
+        int width = originalImage.getWidth();
+        int height = originalImage.getHeight();
+
+        if (width > maxSize || height > maxSize) {
+            double scaleFactor = Math.min((double) maxSize / width, (double) maxSize / height);
+            int targetWidth = (int) (width * scaleFactor);
+            int targetHeight = (int) (height * scaleFactor);
+            finalImage = resizeImage(originalImage, targetWidth, targetHeight);
+        }
+
+        uploadToS3(finalImage, "jpg", metadata, filePath, userId);
+        return filePath;
+    }
+
+    private void uploadToS3(BufferedImage image, String format, ObjectMetadata metadata, String filePath, Long userId) throws IOException {
```
асинхронно это вызывать, вне транзакции
Почему ты в UserService работаешь с IO? Этому тут не место
